File: apps/shared/src/cell/frame-chunks.ts
// Plans and materializes bounded cell-grid snapshot parts from one immutable
// canonical full. Terminal coordinators retain the source and materialize only
// the next cursor part; worker and browser callers may still request every
// deterministic chunk through chunkCellGridFrame.
import { clone, create } from "@bufbuild/protobuf";
import {
  PbCellGridChunkSchema,
  PbCellGridFrameSchema,
  type PbCellGridChunk,
  type PbCellGridFrame,
  type PbCellRow,
} from "../gen/roost/v1/cell_pb.ts";
import { isTerminalUuid } from "../viewport.ts";
import {
  CELL_GRID_PART_MAX_BYTES,
  CELL_GRID_SNAPSHOT_MAX_BYTES,
  CELL_GRID_SNAPSHOT_MAX_CHUNKS,
  assertCellGridSnapshot,
  createCellGridFramePart,
  encodedCellGridChunkSize,
  encodedCellGridFrameSize,
  rejectCellGridChunk,
} from "./frame-chunk-validation.ts";

export {
  CELL_GRID_CHUNK_STALL_MS,
  CELL_GRID_PART_MAX_BYTES,
  CELL_GRID_SNAPSHOT_MAX_BYTES,
  CELL_GRID_SNAPSHOT_MAX_CHUNKS,
  CELL_GRID_SNAPSHOT_MAX_LINK_MAPPINGS,
  CELL_GRID_SNAPSHOT_MAX_ROWS,
  CELL_GRID_SNAPSHOT_MAX_SPANS,
  CellGridChunkError,
  assertCellGridSnapshot,
  decodeCellGridChunk,
  encodeCellGridChunk,
  encodedCellGridChunkSize,
  encodedCellGridFrameSize,
  type CellGridChunkErrorCode,
  type CellGridSnapshotStats,
} from "./frame-chunk-validation.ts";
export {
  CellGridChunkAssembler,
  type CellGridChunkAssemblyResult,
} from "./frame-chunk-assembler.ts";

interface SnapshotRow {
  kind: "history" | "viewport";
  row: PbCellRow;
}

interface PlannedSnapshotPart {
  readonly entries: readonly SnapshotRow[];
  readonly encodedPartBytes: number;
}

export type CellGridSnapshotPart =
  | { readonly kind: "frame"; readonly value: PbCellGridFrame }
  | { readonly kind: "chunk"; readonly value: PbCellGridChunk };

export interface CellGridSnapshotCursor {
  readonly partCount: number;
  materialize(partIndex: number): CellGridSnapshotPart;
}

/**
 * One validated immutable full plus its deterministic part plan. A cursor owns
 * its snapshot UUID, so many recipients can stream the same canonical source.
 */
export interface CellGridSnapshotSource {
  readonly partCount: number;
  createCursor(snapshotId: string): CellGridSnapshotCursor;
}

export interface CreateCellGridSnapshotSourceOptions {
  /** Preserve chunkCellGridFrame's historical one-chunk result for small fulls. */
  readonly forceChunking?: boolean;
}

const PLANNING_SNAPSHOT_ID = "00000000-0000-4000-8000-000000000000";

function encodedVarintBytes(value: number): number {
  let remaining = value;
  let bytes = 1;
  while (remaining >= 0x80) {
    remaining = Math.floor(remaining / 0x80);
    bytes++;
  }
  return bytes;
}

function snapshotRows(frame: PbCellGridFrame): SnapshotRow[] {
  const historyByIndex = new Map(frame.scrollbackRows.map((row) => [row.index, row]));
  const viewportByIndex = new Map(frame.viewportRows.map((row) => [row.index, row]));
  const rows: SnapshotRow[] = [];
  const firstHistoryIndex = Number(frame.sbBase);
  for (let offset = 0; offset < frame.scrollbackRows.length; offset++) {
    const row = historyByIndex.get(firstHistoryIndex + offset);
    if (!row) throw new Error("validated cell snapshot history changed during planning");
    rows.push({ kind: "history", row });
  }
  for (let index = 0; index < frame.rows; index++) {
    const row = viewportByIndex.get(index);
    if (!row) throw new Error("validated cell snapshot viewport changed during planning");
    rows.push({ kind: "viewport", row });
  }
  return rows;
}


function rowContribution(
  frame: PbCellGridFrame,
  emptyPartBytes: number,
  entry: SnapshotRow,
): number {
  const onlyRow = entry.kind === "history"
    ? createCellGridFramePart(frame, [], [entry.row])
    : createCellGridFramePart(frame, [entry.row], []);
  return encodedCellGridFrameSize(onlyRow) - emptyPartBytes;
}

function chunkHeaderBytes(snapshotId: string, chunkIndex: number, chunkCount: number): number {
  const header = create(PbCellGridChunkSchema, { snapshotId, chunkIndex, chunkCount });
  return encodedCellGridChunkSize(header);
}

function partFieldTagBytes(
  emptyFramePart: PbCellGridFrame,
  emptyPartBytes: number,
  headerBytes: number,
): number {
  const encoded = encodedCellGridChunkSize(create(PbCellGridChunkSchema, {
    snapshotId: PLANNING_SNAPSHOT_ID,
    chunkIndex: CELL_GRID_SNAPSHOT_MAX_CHUNKS - 1,
    chunkCount: CELL_GRID_SNAPSHOT_MAX_CHUNKS,
    part: emptyFramePart,
  }));
  return encoded - headerBytes - encodedVarintBytes(emptyPartBytes) - emptyPartBytes;
}

function encodedChunkBytes(
  headerBytes: number,
  partTagBytes: number,
  encodedPartBytes: number,
): number {
  return headerBytes + partTagBytes + encodedVarintBytes(encodedPartBytes) + encodedPartBytes;
}

function planSnapshotParts(frame: PbCellGridFrame): PlannedSnapshotPart[] {
  const basePart = createCellGridFramePart(frame, [], []);
  const basePartBytes = encodedCellGridFrameSize(basePart);
  const largestHeaderBytes = chunkHeaderBytes(
    PLANNING_SNAPSHOT_ID,
    CELL_GRID_SNAPSHOT_MAX_CHUNKS - 1,
    CELL_GRID_SNAPSHOT_MAX_CHUNKS,
  );
  const tagBytes = partFieldTagBytes(basePart, basePartBytes, largestHeaderBytes);
  const parts: PlannedSnapshotPart[] = [];
  let entries: SnapshotRow[] = [];
  let encodedPartBytes = basePartBytes;

  for (const entry of snapshotRows(frame)) {
    const contribution = rowContribution(frame, basePartBytes, entry);
    const nextPartBytes = encodedPartBytes + contribution;
    if (encodedChunkBytes(largestHeaderBytes, tagBytes, nextPartBytes) <= CELL_GRID_PART_MAX_BYTES) {
      entries.push(entry);
      encodedPartBytes = nextPartBytes;
      continue;
    }
    if (entries.length === 0) {
      rejectCellGridChunk(
        "single-row-oversize",
        `cell snapshot ${entry.kind} row ${entry.row.index} cannot fit in ${CELL_GRID_PART_MAX_BYTES} bytes`,
      );
    }
    parts.push({ entries, encodedPartBytes });
    entries = [entry];
    encodedPartBytes = basePartBytes + contribution;
    if (encodedChunkBytes(largestHeaderBytes, tagBytes, encodedPartBytes) > CELL_GRID_PART_MAX_BYTES) {
      rejectCellGridChunk(
        "single-row-oversize",
        `cell snapshot ${entry.kind} row ${entry.row.index} cannot fit in ${CELL_GRID_PART_MAX_BYTES} bytes`,
      );
    }
  }
  if (entries.length > 0) parts.push({ entries, encodedPartBytes });
  if (parts.length === 0 || parts.length > CELL_GRID_SNAPSHOT_MAX_CHUNKS) {
    rejectCellGridChunk(
      "chunk-count",
      `cell snapshot requires ${parts.length} chunks; maximum is ${CELL_GRID_SNAPSHOT_MAX_CHUNKS}`,
    );
  }

  const totalBytes = parts.reduce((total, part, index) => total + encodedChunkBytes(
    chunkHeaderBytes(PLANNING_SNAPSHOT_ID, index, parts.length),
    tagBytes,
    part.encodedPartBytes,
  ), 0);
  if (totalBytes > CELL_GRID_SNAPSHOT_MAX_BYTES) {
    rejectCellGridChunk(
      "snapshot-size",
      `cell snapshot is ${totalBytes} encoded bytes; maximum is ${CELL_GRID_SNAPSHOT_MAX_BYTES}`,
    );
  }
  return parts;
}

class PlannedCellGridSnapshotSource implements CellGridSnapshotSource {
  readonly partCount: number;

  constructor(
    private readonly frame: PbCellGridFrame,
    private readonly parts: readonly PlannedSnapshotPart[] | null,
  ) {
    this.partCount = parts?.length ?? 1;
  }

  createCursor(snapshotId: string): CellGridSnapshotCursor {
    if (this.parts && !isTerminalUuid(snapshotId)) {
      rejectCellGridChunk(
        "invalid-snapshot-id",
        `cell snapshot_id is not a UUID: ${JSON.stringify(snapshotId)}`,
      );
    }
    return {
      partCount: this.partCount,
      materialize: (partIndex) => this.materialize(snapshotId, partIndex),
    };
  }

  private materialize(snapshotId: string, partIndex: number): CellGridSnapshotPart {
    if (!Number.isInteger(partIndex) || partIndex < 0 || partIndex >= this.partCount) {
      rejectCellGridChunk(
        "chunk-index",
        `cell snapshot part index ${partIndex} is outside 0..${this.partCount - 1}`,
      );
    }
    if (!this.parts) {
      return { kind: "frame", value: clone(PbCellGridFrameSchema, this.frame) };
    }
    const plan = this.parts[partIndex]!;
    const viewportRows: PbCellRow[] = [];
    const scrollbackRows: PbCellRow[] = [];
    for (const entry of plan.entries) {
      if (entry.kind === "history") scrollbackRows.push(entry.row);
      else viewportRows.push(entry.row);
    }
    const chunk = create(PbCellGridChunkSchema, {
      snapshotId,
      chunkIndex: partIndex,
      chunkCount: this.partCount,
      part: createCellGridFramePart(this.frame, viewportRows, scrollbackRows),
    });
    const bytes = encodedCellGridChunkSize(chunk);
    if (bytes > CELL_GRID_PART_MAX_BYTES) {
      rejectCellGridChunk("chunk-size", `cell snapshot chunk ${partIndex} is ${bytes} bytes`);
    }
    return { kind: "chunk", value: chunk };
  }
}

/**
 * Validate once, plan whole-row boundaries once, then materialize only a
 * cursor's requested part. The caller must not mutate the canonical frame.
 */
export function createCellGridSnapshotSource(
  frame: PbCellGridFrame,
  options: CreateCellGridSnapshotSourceOptions = {},
): CellGridSnapshotSource {
  assertCellGridSnapshot(frame);
  if (!options.forceChunking && encodedCellGridFrameSize(frame) <= CELL_GRID_PART_MAX_BYTES) {
    return new PlannedCellGridSnapshotSource(frame, null);
  }
  return new PlannedCellGridSnapshotSource(frame, planSnapshotParts(frame));
}

/** Deterministically split a complete full into whole-row bounded chunks. */
export function chunkCellGridFrame(
  frame: PbCellGridFrame,
  snapshotId: string,
): PbCellGridChunk[] {
  const cursor = createCellGridSnapshotSource(frame, { forceChunking: true })
    .createCursor(snapshotId);
  const chunks: PbCellGridChunk[] = [];
  for (let partIndex = 0; partIndex < cursor.partCount; partIndex++) {
    const part = cursor.materialize(partIndex);
    if (part.kind !== "chunk") throw new Error("forced cell snapshot chunk was not chunked");
    chunks.push(part.value);
  }
  return chunks;
}

File: crates/roost-protocol/src/cell/frame_chunks.rs
//! Planning, materialising and sizing the bounded parts of one cell snapshot.
//!
//! A coordinator that retains a canonical full materialises only the part a
//! recipient asks for; a worker or browser may take every deterministic chunk at
//! once. Both go through the one plan computed here, so two recipients of one
//! source always agree on where the row boundaries are. The byte arithmetic
//! that plan is measured with is in `plan`.
//!
//! Every size is buffa's own encoded length: the 1 MiB part limit and the
//! 64 MiB snapshot limit ARE canonical protobuf byte counts, and a second size
//! arithmetic is how one side builds a part the other calls oversized.
mod plan;

use std::fmt;

use roost_proto::buffa::Message;
use roost_proto::{PbCellGridChunk, PbCellGridFrame, PbCellRow};

use self::plan::{PlannedSnapshotPart, plan_snapshot_parts};
use crate::cell::frame_chunk_validation::CELL_GRID_PART_MAX_BYTES;
use crate::cell::frame_chunk_validation::CellGridChunkErrorCode as Code;
use crate::cell::frame_chunk_validation::reject_cell_grid_chunk as reject;
use crate::cell::frame_chunk_validation::{CellGridChunkError, assert_cell_grid_snapshot, quoted};
use crate::viewport::is_terminal_uuid;

/// Canonical encoded size of one frame, in bytes.
pub fn encoded_cell_grid_frame_size(frame: &PbCellGridFrame) -> u32 {
    frame.encoded_len()
}

/// Canonical encoded size of one chunk, in bytes.
pub fn encoded_cell_grid_chunk_size(chunk: &PbCellGridChunk) -> u32 {
    chunk.encoded_len()
}

/// Why a part is over the byte limit.
fn chunk_size_reason(encoded_bytes: u32) -> String {
    let ceiling = CELL_GRID_PART_MAX_BYTES;
    format!("cell snapshot chunk is {encoded_bytes} bytes; maximum is {ceiling}")
}

/// Encode one chunk, refusing anything over the part limit before a byte is made.
pub fn encode_cell_grid_chunk(chunk: &PbCellGridChunk) -> Result<Vec<u8>, CellGridChunkError> {
    let encoded_bytes = encoded_cell_grid_chunk_size(chunk);
    if encoded_bytes > CELL_GRID_PART_MAX_BYTES {
        return Err(reject(Code::ChunkSize, chunk_size_reason(encoded_bytes)));
    }
    Ok(chunk.encode_to_vec())
}

/// A chunk that could not become bytes: refused by the contract, or not protobuf.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CellGridChunkDecodeError {
    Contract(CellGridChunkError),
    Wire(roost_proto::buffa::DecodeError),
}

impl fmt::Display for CellGridChunkDecodeError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Contract(error) => write!(formatter, "{error}"),
            Self::Wire(error) => write!(formatter, "cell snapshot chunk is not protobuf: {error}"),
        }
    }
}

impl std::error::Error for CellGridChunkDecodeError {}

/// Decode one chunk, refusing an oversized buffer before it is parsed.
pub fn decode_cell_grid_chunk(encoded: &[u8]) -> Result<PbCellGridChunk, CellGridChunkDecodeError> {
    if encoded.len() > CELL_GRID_PART_MAX_BYTES as usize {
        let reason = chunk_size_reason(encoded.len() as u32);
        return Err(CellGridChunkDecodeError::Contract(reject(
            Code::ChunkSize,
            reason,
        )));
    }
    PbCellGridChunk::decode_from_slice(encoded).map_err(CellGridChunkDecodeError::Wire)
}

/// One part: the source frame's whole scalar metadata plus exactly the rows
/// this part owns, with unknown fields dropped.
pub fn create_cell_grid_frame_part(
    frame: &PbCellGridFrame,
    rows: &[PbCellRow],
    scrollback_rows: &[PbCellRow],
) -> PbCellGridFrame {
    let mut part = frame.clone();
    part.viewport_rows = rows.to_vec();
    part.scrollback_rows = scrollback_rows.to_vec();
    part.scrollback_append = Vec::new();
    part.__buffa_unknown_fields = Default::default();
    part
}

/// What a cursor hands back for one index: the whole frame, or a chunk.
#[derive(Debug, Clone, PartialEq)]
pub enum CellGridSnapshotPart {
    Frame(PbCellGridFrame),
    Chunk(PbCellGridChunk),
}

/// One validated immutable full plus its part plan. A cursor owns its snapshot
/// UUID, so many recipients can stream the same canonical source.
#[derive(Debug, Clone)]
pub struct CellGridSnapshotSource {
    frame: PbCellGridFrame,
    /// `None` when the whole frame fitted inside one part.
    parts: Option<Vec<PlannedSnapshotPart>>,
}

impl CellGridSnapshotSource {
    /// How many parts every cursor of this source walks.
    pub fn part_count(&self) -> u32 {
        match &self.parts {
            Some(parts) => u32::try_from(parts.len()).unwrap_or(u32::MAX),
            None => 1,
        }
    }

    /// One recipient's view of the source, addressed by its own snapshot UUID.
    pub fn create_cursor(
        &self,
        snapshot_id: &str,
    ) -> Result<CellGridSnapshotCursor<'_>, CellGridChunkError> {
        if self.parts.is_some() && !is_terminal_uuid(snapshot_id) {
            let reason = format!("cell snapshot_id is not a UUID: {}", quoted(snapshot_id));
            return Err(reject(Code::InvalidSnapshotId, reason));
        }
        Ok(CellGridSnapshotCursor {
            source: self,
            snapshot_id: snapshot_id.to_owned(),
            part_count: self.part_count(),
        })
    }
}

/// One recipient's cursor: the same plan, stamped with one snapshot UUID.
#[derive(Debug, Clone)]
pub struct CellGridSnapshotCursor<'a> {
    source: &'a CellGridSnapshotSource,
    snapshot_id: String,
    part_count: u32,
}

impl CellGridSnapshotCursor<'_> {
    pub fn part_count(&self) -> u32 {
        self.part_count
    }

    /// Materialise one part. The plan is fixed, so the same index twice is the
    /// same part.
    pub fn materialize(&self, part_index: u32) -> Result<CellGridSnapshotPart, CellGridChunkError> {
        if part_index >= self.part_count {
            let last = self.part_count - 1;
            let reason = format!("cell snapshot part index {part_index} is outside 0..{last}");
            return Err(reject(Code::ChunkIndex, reason));
        }
        let Some(parts) = &self.source.parts else {
            return Ok(CellGridSnapshotPart::Frame(self.source.frame.clone()));
        };
        let mut viewport_rows = Vec::new();
        let mut scrollback_rows = Vec::new();
        for entry in &parts[part_index as usize].entries {
            let row = entry.row(&self.source.frame)?;
            if entry.history {
                scrollback_rows.push(row.clone());
            } else {
                viewport_rows.push(row.clone());
            }
        }
        let frame = &self.source.frame;
        let part = create_cell_grid_frame_part(frame, &viewport_rows, &scrollback_rows);
        let chunk = PbCellGridChunk {
            snapshot_id: self.snapshot_id.clone(),
            chunk_index: part_index,
            chunk_count: self.part_count,
            part: roost_proto::buffa::MessageField::some(part),
            __buffa_unknown_fields: Default::default(),
        };
        let encoded_bytes = encoded_cell_grid_chunk_size(&chunk);
        if encoded_bytes > CELL_GRID_PART_MAX_BYTES {
            let reason = format!("cell snapshot chunk {part_index} is {encoded_bytes} bytes");
            return Err(reject(Code::ChunkSize, reason));
        }
        Ok(CellGridSnapshotPart::Chunk(chunk))
    }
}

/// Validate once, plan whole-row boundaries once, then materialise only a
/// cursor's part. `force_chunking` keeps a small full as one frame.
pub fn create_cell_grid_snapshot_source(
    frame: &PbCellGridFrame,
    force_chunking: bool,
) -> Result<CellGridSnapshotSource, CellGridChunkError> {
    assert_cell_grid_snapshot(frame)?;
    if !force_chunking && encoded_cell_grid_frame_size(frame) <= CELL_GRID_PART_MAX_BYTES {
        return Ok(CellGridSnapshotSource {
            frame: frame.clone(),
            parts: None,
        });
    }
    let parts = plan_snapshot_parts(frame)?;
    Ok(CellGridSnapshotSource {
        frame: frame.clone(),
        parts: Some(parts),
    })
}

/// Deterministically split a complete full into whole-row bounded chunks.
pub fn chunk_cell_grid_frame(
    frame: &PbCellGridFrame,
    snapshot_id: &str,
) -> Result<Vec<PbCellGridChunk>, CellGridChunkError> {
    let source = create_cell_grid_snapshot_source(frame, true)?;
    let cursor = source.create_cursor(snapshot_id)?;
    let mut chunks = Vec::with_capacity(cursor.part_count() as usize);
    for part_index in 0..cursor.part_count() {
        match cursor.materialize(part_index)? {
            CellGridSnapshotPart::Chunk(chunk) => chunks.push(chunk),
            CellGridSnapshotPart::Frame(_) => {
                let reason = "a forced cell snapshot chunk was not chunked";
                return Err(reject(Code::ChunkCount, reason));
            }
        }
    }
    Ok(chunks)
}
